src/manifold/src/smoothing.cpp

```diff
@@ -501,6 +501,269 @@ class Partition {
+        glm::acos(glm::dot(faceNormal_[e / 3], faceNormal_[pair / 3])));
+    if (dihedral > minSharpAngle) {
+      sharpenedEdges.push_back({e, minSmoothness});
+      sharpenedEdges.push_back({pair, minSmoothness});
+    }
+  }
+  return sharpenedEdges;
+}
+
+/**
+ * Instead of calculating the internal shared normals like CalculateNormals
+ * does, this method fills in vertex properties, unshared across edges that
+ * are bent more than minSharpAngle.
+ */
+void Manifold::Impl::SetNormals(int normalIdx, float minSharpAngle) {
+  if (IsEmpty()) return;
+  if (normalIdx < 0) return;
+
+  const int oldNumProp = NumProp();
+  const int numTri = NumTri();
+
+  Vec<bool> triIsFlatFace = FlatFaces();
+  Vec<int> vertFlatFace = VertFlatFace(triIsFlatFace);
+  Vec<int> vertNumSharp(NumVert(), 0);
+  for (int e = 0; e < halfedge_.size(); ++e) {
+    if (!halfedge_[e].IsForward()) continue;
+    const int pair = halfedge_[e].pairedHalfedge;
+    const int tri1 = e / 3;
+    const int tri2 = pair / 3;
+    const float dihedral =
+        glm::degrees(glm::acos(glm::dot(faceNormal_[tri1], faceNormal_[tri2])));
+    if (dihedral > minSharpAngle) {
+      ++vertNumSharp[halfedge_[e].startVert];
+      ++vertNumSharp[halfedge_[e].endVert];
+    } else {
+      const bool faceSplit =
+          triIsFlatFace[tri1] != triIsFlatFace[tri2] ||
+          (triIsFlatFace[tri1] && triIsFlatFace[tri2] &&
+           !meshRelation_.triRef[tri1].SameFace(meshRelation_.triRef[tri2]));
+      if (vertFlatFace[halfedge_[e].startVert] == -2 && faceSplit) {
+        ++vertNumSharp[halfedge_[e].startVert];
+      }
+      if (vertFlatFace[halfedge_[e].endVert] == -2 && faceSplit) {
+        ++vertNumSharp[halfedge_[e].endVert];
+      }
+    }
+  }
+
+  const int numProp = glm::max(oldNumProp, normalIdx + 3);
+  Vec<float> oldProperties(numProp * NumPropVert(), 0);
+  meshRelation_.properties.swap(oldProperties);
+  meshRelation_.numProp = numProp;
+  if (meshRelation_.triProperties.size() == 0) {
+    meshRelation_.triProperties.resize(numTri);
+    for_each_n(autoPolicy(numTri), countAt(0), numTri, [this](int tri) {
+      for (const int j : {0, 1, 2})
+        this->meshRelation_.triProperties[tri][j] =
+            this->halfedge_[3 * tri + j].startVert;
+    });
+  }
+  Vec<glm::ivec3> oldTriProp(numTri, {-1, -1, -1});
+  meshRelation_.triProperties.swap(oldTriProp);
+
+  for (int tri = 0; tri < numTri; ++tri) {
+    for (const int i : {0, 1, 2}) {
+      if (meshRelation_.triProperties[tri][i] >= 0) continue;
+      int current = 3 * tri + i;
+      const int startEdge = current;
+      const int vert = halfedge_[current].startVert;
+
+      if (vertNumSharp[vert] < 2) {
+        const glm::vec3 normal = vertFlatFace[vert] >= 0
+                                     ? faceNormal_[vertFlatFace[vert]]
+                                     : vertNormal_[vert];
+        int lastProp = -1;
+        do {
+          current = NextHalfedge(halfedge_[current].pairedHalfedge);
```

Comment:
thinking about it, maybe we should just make a halfedge iterator that loops through all the halfedges connected to a vertex, to make this pattern more readable to beginners.

Reply:
It should be possible to make an iterator out of this. I will open an issue about this.